File: Dynamic Programming/1277. Count Square Submatrices with All Ones.js
/**
 * @param {number[][]} matrix
 * @return {number}
 */
var countSquares = function(matrix) {
    const [m, n] = [matrix.length, matrix[0].length];
    let result = 0;
    const dp = []; // dp[i][j] means the number of squres with matrix[i][j] as the bottom-right vertex
    for(let i = 0; i < m; i++) {
        dp[i][j] = [];
        for(let j = 0; j < n; j++) {
           if (i === 0 || j === 0) {
              dp[i][j] = matrix[i][j]; 
           } else if (matrix[i][j] === 0) {
              dp[i][j] = 0;
           }  else {
              dp[i][j] = Math.min(dp[i - 1][j - 1], dp[i - 1][j], dp[i][j - 1]) + 1; 
           }
           result += dp[i][j];
        }
    }
    return result;
};

File: Golang/DynamicProgramming/1277.count-square-submatrices-with-all-ones.go
func countSquares(matrix [][]int) int {
	m, n := len(matrix), len(matrix[0])
	dp := make([][]int, m)
	ret := 0
	for i := range dp {
		dp[i] = make([]int, n)
		dp[i][0] = matrix[i][0]
		ret += dp[i][0]
	}
	for j := 1; j < n; j++ {
		dp[0][j] = matrix[0][j]
		ret += dp[0][j]
	}
	for i := 1; i < m; i++ {
		for j := 1; j < n; j++ {
			if matrix[i][j] == 1 {
				dp[i][j] = min(dp[i-1][j-1], min(dp[i-1][j], dp[i][j-1])) + 1
				ret += dp[i][j]
			}
		}
	}
	return ret
}